### PR TITLE
Various createdump fixes. Smaller MacOS dump size, better logging/stats, misc cleanup, etc.

### LINKED ISSUE #68604: Investigate why createdump generated dumps are so large on MacOS
We are seeing ~6GB dumps on MacOS in the CI jobs.

### PR DESCRIPTION
**Pull request by mikem8361**
Fix where CombineMemoryRegions is called and perf fix for PAGE_SIZE on MacOS M1.

Remove MEMORY_REGION_FLAG_MEMORY_BACKED flags because it was always set now.

Add better memory tracing and memory region stats.

Fix MacOS native module regions when overlapping with existing.

Fix MacOS adding the managed modules to the module mapping list before the "other mappings" is built.

Don't add share_mode == SM_EMPTY regions for MacOS coredumps. Fixes the large M1 dumps.

### CODE REVIEW

**Comment by mikem8361** [comment]
Large MacOS dump issue: https://github.com/dotnet/runtime/issues/68604

File: src/coreclr/debug/createdump/crashinfo.cpp [hunk 1]
```diff
@@ -155,9 +158,28 @@ CrashInfo::GatherCrashInfo(MINIDUMP_TYPE minidumpType)
...
+    }
+    // Enumerate all the managed modules. On MacOS only the native modules have been added
+    // to the module mapping list at this point and adds the managed modules. This needs to
+    // be done before the other mappings is initialized.
+    if (!EnumerateManagedModules())
+    {
+        return false;
+    }
+#ifdef __APPLE__
+    InitializeOtherMappings();
```
Comment (hoyosjs) [inline src/coreclr/debug/createdump/crashinfo.cpp:174]: What other mappings are there here? Other than native mappings and managed ones? Also, why is EnumerateMemoryRegions only called on macOS ?

File: src/coreclr/debug/createdump/crashinfo.cpp [hunk 2]
```diff
@@ -188,9 +210,13 @@ CrashInfo::GatherCrashInfo(MINIDUMP_TYPE minidumpType)
...
             {
                 uint32_t permissions = region.Permissions();
+#ifdef __APPLE__
+                if (permissions == (PF_R | PF_W))
+#else
                 if (permissions == (PF_R | PF_W) || permissions == (PF_R | PF_W | PF_X))
+#endif
                 {
-                    InsertMemoryBackedRegion(region);
+                    InsertMemoryRegion(region);
```
Comment (hoyosjs) [inline src/coreclr/debug/createdump/crashinfo.cpp:219]: Why are we excluding RWX regions on macOS? This should not be all too uncommon?

**Review by hoyosjs — COMMENTED** [review]
Will continue next week

File: src/coreclr/debug/createdump/crashinfo.cpp [hunk 3]
```diff
@@ -457,10 +470,10 @@ CrashInfo::ReplaceModuleMapping(CLRDATA_ADDRESS baseAddress, ULONG64 size, const
         // On MacOS the assemblies are always added.
         MemoryRegion newRegion(flags, start, end, 0, name);
         m_moduleMappings.insert(newRegion);
+        ModuleMappingsCB += newRegion.Size();
```
[the remark this answers is not in the document]
Comment (mikem8361) [reply src/coreclr/debug/createdump/crashinfo.cpp:473]: Yes, that would be a better name for it.

File: src/coreclr/debug/createdump/crashinfo.cpp [hunk 4]
```diff
@@ -636,40 +649,32 @@ CrashInfo::InsertMemoryRegion(uint64_t address, size_t size)
...
     const auto& found = m_memoryRegions.find(region);
     if (found == m_memoryRegions.end())
     {
         // If the region is valid, add the full memory region
-        if (ValidRegion(region)) {
+        if (ValidRegion(region))
+        {
             m_memoryRegions.insert(region);
-            return;
+            return region.Size() / PAGE_SIZE;
```
[the remark this answers is not in the document]
Comment (mikem8361) [reply src/coreclr/debug/createdump/crashinfo.cpp:669]: Adding a MemoryInfo::SizeInPages() method is probably a good idea.  All the pages in this region have been validated and being added as one region.

File: src/coreclr/debug/createdump/crashinfo.cpp (hunk 1 above)
Comment (mikem8361) [reply src/coreclr/debug/createdump/crashinfo.cpp:174]: EnumerateMemoryRegions is the name of the MacOS function that does something similar to EnumerateModuleMappings() on Linux. I can change the name of one of them if that is important.

File: src/coreclr/debug/createdump/crashinfo.cpp (hunk 2 above)
Comment (mikem8361) [reply src/coreclr/debug/createdump/crashinfo.cpp:219]: I'll remove that change. I was thinking that were a lot of them, but there isn't.

Comment (mikem8361) [reply src/coreclr/debug/createdump/crashinfo.cpp:219]: On M1 it does end up being a lot of 16K pages (10MB in my simple test case) so I'm not going to change this back.

Comment (hoyosjs) [reply src/coreclr/debug/createdump/crashinfo.cpp:219]: On m1 R2R code pages are mapped as RWX. Does `MiniDumpWithPrivateReadWriteMemory` imply no module code pages? (I hadn't realized that it was this part of the code).

File: src/coreclr/debug/createdump/crashinfo.cpp [hunk 5]
```diff
@@ -148,7 +148,7 @@ CrashInfo::GatherCrashInfo(MINIDUMP_TYPE minidumpType)
         return false;
     }
     // Gather all the module memory mappings (from /dev/$pid/maps)
-    if (!EnumerateModuleMappings())
+    if (!EnumerateMemoryRegions())
```
Comment (hoyosjs) [inline src/coreclr/debug/createdump/crashinfo.cpp:151]: I meant to ask what was the difference between the calls. Now on apple we are calling EnumerateMemoryRegions twice (140 and this line)

Comment (mikem8361) [reply src/coreclr/debug/createdump/crashinfo.cpp:151]: No, it is only called on line 140. Line 151 is in the #else of the #ifdef __APPLE__

File: src/coreclr/debug/createdump/crashinfo.cpp (hunk 2 above)
Comment (mikem8361) [reply src/coreclr/debug/createdump/crashinfo.cpp:219]: Yes, it means just rw heap pages.  Most code pages are not part of a heap dump with a couple of exceptions.

**Comment by mikem8361** [comment]
I'm hoping we can get this approved for Preview 7 before July 12th.

File: src/coreclr/debug/createdump/createdumpunix.cpp [hunk 6]
```diff
@@ -14,6 +18,12 @@ CreateDump(const char* dumpPathTemplate, int pid, const char* dumpType, MINIDUMP
     std::string dumpPath;
     bool result = false;
 
+    // Initialize PAGE_SIZE
+#if !defined(PAGE_SIZE) && (defined(__arm__) || defined(__aarch64__) || defined(__loongarch64))
+    g_pageSize = sysconf(_SC_PAGESIZE);
+#endif
+    TRACE("PAGE_SIZE %d\n", PAGE_SIZE);
```
[the remark this answers is not in the document]
Comment (mikem8361) [reply src/coreclr/debug/createdump/createdumpunix.cpp:25]: Yes, that makes sense though it does build/work on MacOS M1 (PAGE_SIZE must be defined outside of createdump). I'll remove the !defined(PAGE_SIZE).

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Comment by danmoseley** [comment]
@janvorli's original comment in #68604

> Just FYI, I am getting large dump sizes (over 3GB at least) even for a trivial C app with main containing only a dereference of NULL and nothing else. 

What is the size of the dump from such an experiment now? I am wondering whether we now have dumps comparable in size to what we get on Linux. 

Also I recall compression was hypothesized but if you've now eliminated all the blank pages presumably that's not helpful/not needed.

**Comment by mikem8361** [comment]
MacOS dump sizes are at least 50% to 75% smaller. For the same smallest test app, the MacOS and Linux sizes are comparable.

**Comment by danmoseley** [comment]
That's great to hear. I'm not sure what the state of dump gathering on helix for Mac is right now @MattGal  is it switched on and working but size capped or something?

**Comment by tommcdon** [comment]
> That's great to hear. I'm not sure what the state of dump gathering on helix for Mac is right now @MattGal is it switched on and working but size capped or something?

@MattGal I believe helix is using os dumps (enabled via ulimit).  We should consider switch it to use createdump so we can take advantage of the reduced size.  This will require setting the correct environment variables to enable coreclr-generated crash dumps: https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dumps#collect-dumps-on-crash.

**Comment by MattGal** [comment]
> That's great to hear. I'm not sure what the state of dump gathering on helix for Mac is right now @MattGal is it switched on and working but size capped or something?

Correct.  On-premises machines, due to the more limited network connection, are set to upload up to 2 GB of dumps and a max of 1.5 GB per dump.  As always if you see issues where you think one should have been uploaded but wasn't, we can investigate.

**Comment by hoyosjs** [comment]
@MattGal All macOS system dumps are huge. Have we ever had a successful upload? If not, then we might as well disable the system support and use the environment variables to get the small chunk of actionable dumps. Or we could try to compress them (we realized a lot of pages are zeros, so they are easily compressible).

**Comment by hoyosjs** [comment]
> What do you recommend? We change the dotnet/runtime code above to un-set ulimit and set createdump iff it's not WASM, NativeAOT and Mono?

 Yeah, I am looking at doing this to at least get a section of actionable data.

**Comment by danmoseley** [comment]
> we realized a lot of pages are zeros, so they are easily compressible

That was'nt solved by this change to skip SM_EMPTY ?

**Comment by hoyosjs** [comment]
I mean for system dumps. Createdump dumps are not reasonably smaller

**Comment by danmoseley** [comment]
Is there a way that createdump could terminate in such a way that system dumps were not triggered? this seems arguably a customer scenario -- it's messy if when createdump is triggered JIT it always leads to two dumps on many machines?

**Comment by hoyosjs** [comment]
> Is there a way that createdump could terminate in such a way that system dumps were not triggered? 

Haven't been able to find such a provision for macOS, and it's a fair amount of work for Linux I believe (I'd have to check how apport does it). 

> this seems arguably a customer scenario -- it's messy if when createdump is triggered JIT it always leads to two dumps on many machines?

It sort of is to some extent - the user configured both to be collected though. On top of that, system dumps are fairly hard to use for managed code as they don't have information to map native to managed threads. the createdump dumps can deal with native crashes after PAL initialization, so I'd say they give us a lot better UX for crash diagnostics and solves a lot of our size problems for uploads.

**Comment by hoyosjs** [comment]
Hmm, thinking about this a bit more. I'll need to take a look at the full dumps createdump generates. There are cases where full dumps are necessary for runtime bugs, but I don't know if they are that much lighter than system dumps. @jkotas, do you know how frequently a full dump is needed over a heap dump in runtime bugs? Right now we really get no dumps, so a heap dump is better than nothing, I want to see how much worse off we'd be.

**Comment by jkotas** [comment]
Good percentage of crashes can be diagnosed from heap dumps. I agree that having some dumps is better than no dumps.

**Comment by danmoseley** [comment]
If the issue is that enabling createdump would mean 2 dumps, and that's too much, and it's not easy to prevent it, and the alternative is nothing at all, can we set the script so that if a createdump dump is present (identified using DOTNET_DbgMiniDumpName pattern) it only uploads the createdump dump, otherwise it uploads the system dump?

**Comment by danmoseley** [comment]
In fact, if that scheme is reasonable, it could be done at the Arcade level for everyone?

**Comment by hoyosjs** [comment]
Might have some collisions on PID, though may be more rare than I think. Also, if a process enables one I'd disable the other - they tend to fill the disks fast and make it really hard for subsequent tests to do much. I'd say lets experiment with runtime where we can revise policy quickly and then move the change to arcade (I am not sure helix should always set the env vars during provisioning, but the helix SDK might be a good place to add the support).

**Comment by danmoseley** [comment]
Sounds good - do you plan to take care of that @hoyosjs ?

**Comment by hoyosjs** [comment]
Won't get to this right now, but I've self assigned this. I'll investigate a bit on dump size and determine what the best approach is for runtime. Then I'll sync with Matt who has more context on the Helix upload policy.

**Comment by danmoseley** [comment]
Sounds great, thanks @hoyosjs .

**Pull request merged**